File: 152.乘积最大子数组.cpp
/*
 * @lc app=leetcode.cn id=152 lang=cpp
 *
 * [152] 乘积最大子数组
 */

// @lc code=start
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        int dp[n][2];
        dp[0][0] = nums[0];
        dp[0][1] = nums[0];
        int ans = nums[0];
        for (int i = 1; i < n; i++) {
            if (nums[i] >= 0) {
                dp[i][0] = max(nums[i], dp[i-1][0] * nums[i]);
                dp[i][1] = min(nums[i], dp[i-1][1] * nums[i]);
            }
            else {
                dp[i][0] = max(nums[i], dp[i-1][1] * nums[i]);
                dp[i][1] = min(nums[i], dp[i-1][0] * nums[i]);
            }
            ans = max(ans, dp[i][0]);
        }
        return ans;
    }
};
// @lc code=end

File: 152.乘积最大子数组.py
# @before-stub-for-debug-begin
from python3problem152 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode.cn id=152 lang=python3
#
# [152] 乘积最大子数组
#

# @lc code=start
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max1, max2, maxTotal = 0, 0, 0, 
        for num in nums:
            max1, max2 = max(max1 * num, num) if num >= 0 else max2 * num, max2 * num if num >= 0 else min(max1*num, num)
            maxTotal = max(maxTotal, max1)
        return maxTotal if maxTotal > 0 else nums[0] if len(nums) == 1 else 0
# @lc code=end

